Added 'invert' option to boolean-like bindings

kylefarris opened the pull request:
I often find the need to use the 'toggle' binding but find myself needing the opposite effect. So, for instance, if I have a binding like this:

```javascript
'model.comments.length': {
    type: 'toggle',
    hook: 'no-comments-msg'
}
```

It would be great if I could expect the boolean interpretation of the value to have the opposite effect on the visibility of the element represented by the hook/selector. In other words, in this case, when there are no comments (array is length 0), I want to _show_ a message.

I added a new property to the `ampersand-dom-bindings` **toggle** type called `opposite` that will allow the value of the element to be interpreted in its negated form. So, if the value is usually interpreted to be true, it will now be false and vice versa.

**Example:**

```javascript
'model.comments.length': {
    type: 'toggle',
    opposite: true,
    hook: 'no-comments-msg'
}
```

It's a pretty simple change overall that will add a lot of convenience. Before this, I've had to create derived properties in my models like 'has_no_comments' just to have this effect.

-Kyle

wraithgar commented:
First off, thanks for making this PR. The failed CI tests are not your fault, we're still trying to iron out some kinks in our CI testing.

I think this could be a valuable addition. In the past i've had to make derived properties that toggle the values of booleans to get this behavior.

Would you be willing to add this option to the other boolean-like bindings?  It feels like a natural extension of the idea to make it work on `booleanClass` and `booleanAttribute`.

Finally, the word `opposite` jumps out as confusing at first, but that could be just me.  If no better alternatives are quickly suggested it's not a deal breaker imo.

wraithgar commented:
Actually I spoke too soon, it looks like the tests are failing genuinely for this PR.

kylefarris commented:
Yeah, that's totally my fault. I actually just edited the code directly in github's editor since it was such a small amount of work. Of course, that doesn't trigger the commit hooks or allow to me run the tests myself before they hit Travis. In any case, I cloned the fork on my machine, fixed the test, added the feature to the additional APIs and wrote tests/updated docs for them. Travis is happy now. Yay!

As for naming the property, it doesn't matter to me, 'opposite' was just the first thing that came to mind. If we wanted to keep it a boolean, `reverse` could work. Another option would be where the user specifies how they want the value interpreted. Like 'opposite' or 'normal'. I suppose we could even create a way to allow them to provide a custom interpretation function that has to return either a `true` or `false` value; however, I feel like this is now getting into `switch` territory.

**Examples:**

```javascript
'model.property': {
   type: 'toggle',
   hook: 'foobar',
   interpret: 'opposite' // 'normal' would be the default
}
```
```javascript
'model.property': {
   type: 'toggle',
   hook: 'foobar',
   interpret: function(val) { if (val === 27) return true; return false; }
}
```

Just some ideas.

cdaringe commented:
@kylefarris, i have well have done this manually via functions.  thanks for taking the time!

> ... the word opposite jumps out as confusing at first ...

i was kind of thinking the same thing, but would def take it over nothing.  something a little more common in the dev vernacular may be `invert`?  how do folks feel about that?

kylefarris commented:
I'm totally open to changing it to whatever consensus we come to--makes no difference to me. `invert` could work well and I like that it's a verb instead of an adjective.

@wraithgar any opinion on the matter?

wraithgar commented:
Invert definitely seems better at first glance than opposite.  If you'd be willing to change it to that, great.

The PR looks good and I can see this being a very useful addition that lots of people will use. Thank you!

cdaringe commented:
ditto.  big thanks!

kylefarris commented:
Alright, the configuration property is now called `invert`. Updated tests and README accordingly.

wraithgar commented:
Yes! +1 this is a very nice PR thank you!

kylefarris commented:
My pleasure.

cdaringe commented:
+1

wraithgar commented:
Published as `v3.8.0`
